agent/internal/fluent.go

```diff
@@ -0,0 +1,439 @@
+			return errors.Wrap(err, "failed to kill existing Fluent Bit container")
+		}
+	}
+	return nil
+}
+
+func pullImageByName(ctx *actor.Context, c *client.Client, imageName string) error {
+	_, _, err := c.ImageInspectWithRaw(context.Background(), imageName)
+	switch {
+	case err == nil:
+		// No error means the image is present; do nothing.
+	case client.IsErrNotFound(err):
+		// This error means the call to Docker went fine but the image doesn't exist; pull it now.
+		ctx.Log().Infof("pulling Docker image %s", imageName)
+		pullResponse, pErr := c.ImagePull(context.Background(), imageName, types.ImagePullOptions{})
+		if pErr != nil {
+			return pErr
+		}
+		if _, pErr = ioutil.ReadAll(pullResponse); pErr != nil {
+			return pErr
+		}
+		if pErr = pullResponse.Close(); pErr != nil {
+			return pErr
+		}
+	default:
+		// Something unexpected happened; propagate the error.
+		return err
+	}
+	return nil
+}
+
+// startLoggingContainer starts a Fluent Bit container. It returns the host and port that the agent
+// should use to connect to the daemon, the exposed host port of the container, and the ID of the
+// container.
+func startLoggingContainer(ctx *actor.Context, opts Options) (string, int, int, string, error) {
+	c, err := client.NewClientWithOpts(client.FromEnv, client.WithVersion("1.40"))
+	if err != nil {
+		return "", 0, 0, "", errors.Wrap(err, "error connecting to Docker daemon")
+	}
+
+	defer func() {
+		if cErr := c.Close(); cErr != nil {
+			logrus.WithError(cErr).Warn("failed to close Docker connection")
+		}
+	}()
+
+	const containerName = "determined-fluent"
+	imageName := opts.FluentLoggingImage
+	exposedPort := nat.Port(fmt.Sprintf("%d/tcp", fluentListenPort))
+
+	if err = killContainerByName(ctx, c, containerName); err != nil {
+		return "", 0, 0, "", errors.Wrap(err, "failed to kill old logging container")
+	}
+
+	if err = pullImageByName(ctx, c, imageName); err != nil {
+		return "", 0, 0, "", errors.Wrap(err, "failed to pull logging image")
+	}
+
+	fluentArgs, fluentFiles, err := fluentConfig(opts)
+	if err != nil {
+		return "", 0, 0, "", errors.Wrap(err, "failed to configure Fluent Bit")
+	}
+
+	// Decide what Docker network to use for Fluent Bit. If this agent is itself running inside Docker,
+	// use one of the networks that this container is using. If not, use the default network
+	// ("bridge").
+	dockerNet, err := getDockerNetwork(c)
+	if err != nil {
+		return "", 0, 0, "", errors.Wrap(err, "failed to get Docker network")
+	}
+	fluentDockerNet := dockerNet
+	if fluentDockerNet == "" {
+		fluentDockerNet = "bridge"
+	}
+	ctx.Log().Infof("running Fluent Bit on Docker network %q", fluentDockerNet)
+
+	createResponse, err := c.ContainerCreate(
+		context.Background(),
+		&container.Config{
+			Image:        imageName,
+			Cmd:          fluentArgs,
+			ExposedPorts: nat.PortSet{exposedPort: struct{}{}},
+		},
+		&container.HostConfig{
+			AutoRemove:  true,
```

Comment:
non-blocking: I think we should also start the container with some reasonable limits for `HostConfig.Resources`